File: UtilitiesVB/Base Functionality Extensions/MergeSortImplementations.vb
Public Module MergeSortImplementations
    Private Function MergeSort(Of T)(ByVal coll As IList(Of T), ByVal comparison As Comparison(Of T)) As IList(Of T)
        'DONE: 2023-03-06 Move To UtilitiesVB. A generic merge sort(of T) doesn't belong in a form controller
        Dim Result As New List(Of T)()
        Dim Left As New Queue(Of T)()
        Dim Right As New Queue(Of T)()
        If coll.Count <= 1 Then Return coll
        Dim midpoint As Integer = coll.Count / 2

        For i As Integer = 0 To midpoint - 1
            Left.Enqueue(coll(i))
        Next

        For i As Integer = midpoint To coll.Count - 1
            Right.Enqueue(coll(i))
        Next


        Left = New Queue(Of T)(MergeSort(Left.ToList(), comparison))
        Right = New Queue(Of T)(MergeSort(Right.ToList(), comparison))
        Result = Merge(Left, Right, comparison)
        Return Result
    End Function


    Private Function Merge(Of T)(ByVal Left As Queue(Of T), ByVal Right As Queue(Of T), ByVal comparison As Comparison(Of T)) As List(Of T)
        'DONE: 2023-03-06 Move To UtilitiesVB. A generic merge sort(of T) doesn't belong in a form controller

        Dim Result As New List(Of T)()

        While Left.Count > 0 AndAlso Right.Count > 0
            Dim cmp As Integer = comparison(Left.Peek(), Right.Peek())
            If cmp < 0 Then
                Result.Add(Left.Dequeue())
            Else
                Result.Add(Right.Dequeue())
            End If
        End While

        While Left.Count > 0
            Result.Add(Left.Dequeue())
        End While

        While Right.Count > 0
            Result.Add(Right.Dequeue())
        End While

        Return Result
    End Function
End Module

File: UtilitiesCS/HelperClasses/MergeSortImplementations.cs
using Deedle.Internal;
using System;
using System.Collections.Generic;
using System.Linq;

namespace UtilitiesCS
{
    public static class MergeSortImplementations
    {

        /// <summary>
        /// 2 Overloads
        /// <para>Generic implementation of merge sort algorithm. </para>
        /// <para> Parameter <c><paramref name="inplace"/></c>
        /// determines whether the method returns a new <seealso cref="IList{T}"/> or copies the results 
        /// back into the original parameter named <c><paramref name="list"/></c>.</para>
        /// </summary>
        /// <typeparam name="T">Generic type parameter</typeparam>
        /// <param name="list">the <seealso cref="IList{T}"/> to be sorted </param>
        /// <param name="comparison">Represents the method that compares two objects of the same type.</param>
        /// <param name="inplace"><seealso cref="bool"/> to determine whether to copy back into the original variable</param>
        /// <returns><code>null | <seealso cref="IList{T}"/></code> 
        /// <br/>
        /// <c>null</c> - when <c>InPlace</c> is <c>true</c>. Values are copied to <paramref name="list"/>
        /// <br/>
        /// <seealso cref="IList{T}"/> - when <c>InPlace</c> is <c>false</c>
        /// </returns>
        public static IList<T> MergeSort<T>(this IList<T> list, Comparison<T> comparison, bool inplace)
        {
            var count = list.Count;
            var result = list.MergeSort(comparison);

            if (inplace)
            {
                for (int i = 0; i < count; i++)
                {
                    list[i] = result[i];
                }

                return null;
            }
            
            return result; 
        }

        /// <summary>
        /// Generic implementation of merge sort algorithm
        /// </summary>
        /// <typeparam name="T">Generic type parameter</typeparam>
        /// <param name="list">the <seealso cref="IList{T}"/> to be sorted </param>
        /// <param name="comparison">Represents the method that compares two objects of the same type.</param>
        /// <returns>The sorted <seealso cref="IList{T}"/></returns>
        public static IList<T> MergeSort<T>(this IList<T> list, Comparison<T> comparison)
        {
            var left = new Queue<T>();
            var right = new Queue<T>();
            var count = list.Count;
            if (count <= 1)
                return list;
            int midpoint = (int)Math.Round(count / 2d);

            for (int i = 0, loopTo = midpoint - 1; i <= loopTo; i++)
                left.Enqueue(list[i]);

            for (int i = midpoint, loopTo = count - 1; i <= loopTo; i++)
                right.Enqueue(list[i]);

            left = new Queue<T>(MergeSort(left.ToList(), comparison));
            right = new Queue<T>(MergeSort(right.ToList(), comparison));
            var result = left.Merge(right, comparison);

            for (int i = 0; i< count; i++)
            {
                list[i] = result[i];
            }

            return list;
        }

        private static List<T> Merge<T>(this Queue<T> left, Queue<T> right, Comparison<T> comparison)
        {
            var result = new List<T>();

            while (left.Count > 0 && right.Count > 0)
            {
                int comp = comparison(left.Peek(), right.Peek());
                if (comp < 0) { result.Add(left.Dequeue()); }
                else { result.Add(right.Dequeue()); }
            }

            while (left.Count > 0)
                result.Add(left.Dequeue());

            while (right.Count > 0)
                result.Add(right.Dequeue());

            return result;
        }

    }
}